codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/UnconstrainedShapeSymbolProvider.kt

```diff
@@ -75,22 +79,42 @@ import software.amazon.smithy.rust.codegen.server.smithy.generators.serverBuilde
 class UnconstrainedShapeSymbolProvider(
     private val base: RustSymbolProvider,
     private val model: Model,
-    private val serviceShape: ServiceShape,
     private val publicConstrainedTypes: Boolean,
+    private val serviceShape: ServiceShape,
 ) : WrappingSymbolProvider(base) {
     private val nullableIndex = NullableIndex.of(model)
+    private val visibility = when (publicConstrainedTypes) {
+        true -> Visibility.PUBLIC
+        false -> Visibility.PUBCRATE
+    }
+
+    /**
+     * Unconstrained type names are always suffixed with `Unconstrained` for clarity, even though we could dispense with it
+     * given that they all live inside the `unconstrained` module, so they don't collide with the constrained types.
+     */
+    private fun unconstrainedTypeNameForCollectionOrMapOrUnionShape(shape: Shape): String {
+        check(shape is CollectionShape || shape is MapShape || shape is UnionShape)
+        return RustReservedWords.escapeIfNeeded(shape.contextName(serviceShape).toPascalCase() + "Unconstrained")
+    }
 
+    // this is required because the `Name` field on the symbol is `Vec` (since usually it would be rendered as `Vec<Foo>`
+    // directly
     private fun unconstrainedSymbolForCollectionOrMapOrUnionShape(shape: Shape): Symbol {
         check(shape is CollectionShape || shape is MapShape || shape is UnionShape)
 
-        val name = unconstrainedTypeNameForCollectionOrMapOrUnionShape(shape, serviceShape)
-        val namespace = "crate::${Unconstrained.namespace}::${RustReservedWords.escapeIfNeeded(name.toSnakeCase())}"
-        val rustType = RustType.Opaque(name, namespace)
+        val name = unconstrainedTypeNameForCollectionOrMapOrUnionShape(shape)
+        val module = RustModule.new(
+            RustReservedWords.escapeIfNeeded(name.toSnakeCase()),
+            // NOTE: this is to preserve the codegen diff, but pub(crate) modules are the same as private ones
```

Comment:
```suggestion
```

This is not true. These are inline modules _within_ the `unconstrained` module. They have to be `pub(crate)` for code within e.g. the `json_deser` module to reach them.